### PR TITLE
wallet2_basic: robust compat lib for loading/storing legacy wallets

### PR DESCRIPTION
**Pull request by jeffro256**
This library has no dependency on wallet2.h and gives us a way forward to move away from `wallet2` in the (not-so-distant) future, while still supporting conversions of old wallet files. This lib is also useful if you have an application where you want to extract information directly from the wallet file with or without having to setup accounts and devices. This is now possible because I have split the wallet keys loading into two steps: `load_from_memory` and `setup_account_and_devices`. When one is loading a wallet keys file, the user of the API can choose whether or not to contact external devices during this process with use of the flag `allow_external_devices_setup`.

Reorganized for `seraphis_migration`.

### CODE REVIEW

**Comment by rbrunner7** [comment]
Just to be sure and avoid misunderstandings: This is meant to be "the" commit regarding this functionality and code, right? If reviewed and merged, it would become part of the developing Seraphis enabled Monero prototype, and further development would happen here as well if needed, with further PRs to the same repo?

**Comment by jeffro256** [comment]
Yes, that's the plan currently. I'll probably just add comments, tests, etc

**Comment by rbrunner7** [comment]
Little comment: To successfully make a *debug* build of this code, a small correction is needed that @UkoeHB made in [this commit](https://github.com/UkoeHB/monero/commit/7e9203facfe807a7bee37162aecda67de2c0d590).

**Comment by rbrunner7** [comment]
@jeffro256: I think it would be very useful to add a high-level method to write to file because I believe it would make testing this much easier. The original PR where `wallet2` used this code to write to file did of course not need it, and offered an easy way to test, but this code here does not.

**Comment by jeffro256** [comment]
> Little comment: To successfully make a _debug_ build of this code, a small correction is needed that @UkoeHB made in [this commit](https://github.com/UkoeHB/monero/commit/7e9203facfe807a7bee37162aecda67de2c0d590).

That should be fixed by merging those changes in a different commit/PR.

File: src/wallet/wallet2_basic/wallet2_storage.cpp
Comment (jeffro256) [reply src/wallet/wallet2_basic/wallet2_storage.cpp:259]: Hmm... that seems to be a bug in the compiler. The `[[noreturn]]` annotations on `throw_wallet_ex()` should indicate to the compiler that the control flow ends there. However, I could modify it so the compiler can't mess it up.

**Comment by rbrunner7** [comment]
> > Little comment: To successfully make a _debug_ build of this code, a small correction is needed that @UkoeHB made in [this commit](https://github.com/UkoeHB/monero/commit/7e9203facfe807a7bee37162aecda67de2c0d590).
> 
> That should be fixed by merging those changes in a different commit/PR.

I just saw that I merged the change in question already into our repository, resulting in [this](https://github.com/seraphis-migration/monero/commit/7e9203facfe807a7bee37162aecda67de2c0d590). So I think if you bring your code to this base the problem is already solved.

**Comment by jeffro256** [comment]
> > > Little comment: To successfully make a _debug_ build of this code, a small correction is needed that @UkoeHB made in [this commit](https://github.com/UkoeHB/monero/commit/7e9203facfe807a7bee37162aecda67de2c0d590).
> > 
> > 
> > That should be fixed by merging those changes in a different commit/PR.
> 
> I just saw that I merged the change in question already into our repository, resulting in [this](https://github.com/seraphis-migration/monero/commit/7e9203facfe807a7bee37162aecda67de2c0d590). So I think if you bring your code to this base the problem is already solved.

That commit was merged into `seraphis_lib`, but also needs to be merged into branch `seraphis_wallet`, since `seraphis_wallet` is behind `seraphis_lib`.

**Comment by rbrunner7** [comment]
Duh, yes of course. Separate branches, we don't *directly* use the Seraphis lib branch ...

**Comment by rbrunner7** [comment]
I finally started to test this today.

By pure chance I stumbled over something that may be a not yet supported case: In the CLI wallet, there is a setting available called `export-format`, with the two possible values `binary` and `ascii`. If you have `ascii` active, as I actually had in the very first wallet I took for testing, surprisingly the setting influences the .keys of a wallet as well, which is not binary, but textual, with a first line of:

    -----BEGIN MoneroAsciiDataV1-----

Maybe you encountered the code to handle this while converting it to `wallet2_basic` and decided to leave it out and make this edge case unsupported?

**Comment by rbrunner7** [comment]
I tried with a (not yet finalized) multisig wallet and got the following error:

    Error: Field multisig_derivations not found in JSON

That's expected, right? I think I saw a multisig-related "TODO" somewhere in the code.

Any plans yet when you will go back to those TODOs that are still there?

**Comment by rbrunner7** [comment]
Just FYI: There was a first review of basically this code a while back by @vtnerd, as it started life as [this Monero main code PR](https://github.com/monero-project/monero/pull/8923). Reviewing this here therefore does not start from zero, so to say.

**Comment by vtnerd** [comment]
Why not merge this into monero-project? The more this repo deviates from the main repo, the more painful it will be to merge it all back.

**Comment by rbrunner7** [comment]
> Why not merge this into monero-project? The more this repo deviates from the main repo, the more painful it will be to merge it all back.

I am not sure how good your overview is already regarding changes that will be needed to implement Jamtis and Seraphis. Current thinking is to make `wallet2` private within the codebase, not available anymore to any wallet apps, as an especially striking example. Thus however the Monero master code and the new Seraphis based code finally find together again, it will be hard to reconcile everything, pretty much whatever we do.

That's why personally I don't worry too much about this code here in particular. It's all new, so merging back shouldn't be a big problem. Keeping it aligned with structure changes in wallet files over the next 2, maybe 3, years should be manageable.

On the other hand we can move quite a bit faster here with "our" repository as an advantage, IMHO.

But well, that's just my opinion; interested what other people think.

**Comment by rbrunner7** [comment]
@DangerousFreedom1984: I also think that this can and should be merged soon, even if there as still some gaps in the functionality. I tested it today, it works solidly. I also looked over the code, ok as well. Further improvements are possible without much fuss with further PRs.

Strong opinions that this PR goes to the wrong repository altogether notwithstanding of course, see my take on @vtnerd 's question above :)

**Comment by rbrunner7** [comment]
@jeffro256 : This code still contains some TODOs. What would you propose: You do those before we merge this, or we merge now and you make later separate PRs to do those?

I would be ready to merge now, to have the code easily available, but wonder how we would make sure those TODOs do not fall by the wayside.

And another question: With the code as of today, is there anything that could get lost if I load and then save a wallet because some data is not supported? E.g. a multisig wallet will not lose everythig multisig related if I load it and save it with that code?

**Comment by jeffro256** [comment]
The TODOs in the unit tests are there because there's a couple of fields which aren't present in that specific wallet file, so those specific fields aren't tested. The TODOs in the key loading/storing are just me saying the code could be more conservative and assert the variant values of enums, but it isn't needed unless the user manually modifies those fields to be an invalid enum value. 

> And another question: With the code as of today, is there anything that could get lost if I load and then save a wallet because some data is not supported? E.g. a multisig wallet will not lose everythig multisig related if I load it and save it with that code?

It shouldn't... I tried to specifically make it a 1-to-1 matching for `wallet2` files. However, if you do find a bug that removed information while storing, please let me know ;)

**Comment by rbrunner7** [comment]
Thanks for the info, @jeffro256 . So from my side there probably only is [this observation](https://github.com/seraphis-migration/monero/pull/4#issuecomment-1732586237) left to check. Maybe I will debug to have a look if it's not immediately clear to you where the problem is.

**Comment by jeffro256** [comment]
> I tried with a (not yet finalized) multisig wallet and got the following error:

That's a great find! That field shouldn't be mandatory for all multisig wallets

**Comment by jeffro256** [comment]
> Why not merge this into monero-project? The more this repo deviates from the main repo, the more painful it will be to merge it all back.


@vtnerd I'll honestly be convinced either way, but currently I'm of the mind that since this will be the most useful here, and since it may be iterated upon, we can mold it here first then pass on a more mature version to the core repo

File: src/wallet/wallet2_basic/wallet2_storage.cpp
Comment (rbrunner7) [inline src/wallet/wallet2_basic/wallet2_storage.cpp:625]: I think now I found a small thing that is not implemented. As reported [here](https://github.com/seraphis-migration/monero/pull/4#issuecomment-1731384481) it looked as if loading a .keys file in ASCII format is not yet supported. Now, looking at your code, I think I confirmed that, because support for `ExportFormat::Ascii` is nowhere to be found.

Comment (jeffro256) [reply src/wallet/wallet2_basic/wallet2_storage.cpp:672]: Okay, the wallet now supports loading the keys using PEM ascii format, automatically detecting it.

Comment (jeffro256) [reply src/wallet/wallet2_basic/wallet2_storage.cpp:672]: (and storing it)

**Comment by jeffro256** [comment]
Only kink I'm working out now is how to load/store with non-standard KDF rounds

**Comment by jeffro256** [comment]
Okay, I fixed the multiple KDF rounds issue. Is there anything loading/storing schemes that I could support?

**Comment by rbrunner7** [comment]
>  Is there any other loading/storing schemes that I could support?

I can't think of any. I had no wallet older than 2017 to test, but I don't see why it would fail with even older ones, with the code being a 1-to-1 extract of the corresponding `wallet2` code.

And anyway, with such things the only way to really test is using, using, using :)

If nobody opposes I intend to merge this PR tomorrow.

**Review by rbrunner7 — APPROVED** [review]
Had a final look over the code. I did not systematically compare this code with the code parts in `wallet2` where it comes from, but I really don't think that's necessary for this review: If the code works, i.e. loads and saves wallets correctly, we can be pretty sure the transformation was correct.

**Pull request merged**
